packages/medusa/src/api/utils/validate-body.ts

```diff
@@ -4,15 +4,29 @@ import { MedusaRequest, MedusaResponse } from "../../types/routing"
 import { zodValidator } from "./zod-helper"
 
 export function validateAndTransformBody(
-  zodSchema: z.ZodObject<any, any> | z.ZodEffects<any, any>
+  zodSchema:
+    | z.ZodObject<any, any>
+    | ((
+        customSchema?: z.ZodObject<any, any>
+      ) => z.ZodObject<any, any> | z.ZodEffects<any, any>)
 ): (
   req: MedusaRequest,
   res: MedusaResponse,
   next: NextFunction
 ) => Promise<void> {
   return async (req: MedusaRequest, _: MedusaResponse, next: NextFunction) => {
     try {
-      req.validatedBody = await zodValidator(zodSchema, req.body)
+      let schema: z.ZodObject<any, any> | z.ZodEffects<any, any>
+
+      if (typeof zodSchema === "function") {
+        schema = zodSchema(req.bodyValidator)
```

Comment:
req.bodyValidator can be undefined right?

Reply:
Yes it can be!